utils/land-patterns.stanza

```diff
@@ -1885,3 +1907,275 @@ public defn make-two-pin-axial-landpattern (length:Toleranced,
+;====================================== BGA Land Patterns =========================================
+;==================================================================================================
+public pcb-enum ocdb/land-patterns/BGAPadType:
+  SolderMaskDefined,
+  NonSolderMaskDefined,
+  Collapsible,
+  
+pcb-pad bga-pad (typical-ball-diameter:Double,      ; the typical/nominal size of the bga balls
+                 pad-type:BGAPadType,               ; the bga pad type, (collapsible, solder mask defined, or non-soldermask defined)
+                 solder-mask-registration:Double,   ; the solder-mask registration of the design
+                 density-level:DensityLevel,        ; the density level of the design
+                 ):
+  name = "BGA Pad"
+  type = SMD
+  val solder-mask-defined? = pad-type is SolderMaskDefined
+  val collapsible? = pad-type is Collapsible
+
+  val adjustment = switch(density-level):
+    DensityLevelA: 0.75 when collapsible? else 1.15
+    DensityLevelB: 0.80 when collapsible? else 1.10
+    DensityLevelC: 0.85 when collapsible? else 1.05
+
+  val definition-diameter = typical-ball-diameter * adjustment
+  val referenced-diameter = definition-diameter + solder-mask-registration * 2.0
+
+  if solder-mask-defined?:
+    shape = Circle(0.5 * definition-diameter)
+    layer(SolderMask(Top)) = Circle(0.5 * referenced-diameter)
+  else:
+    shape = Circle(0.5 * referenced-diameter)
+    layer(SolderMask(Top)) = Circle(0.5 * definition-diameter)
+
+defn bga-pad (typical-ball-diameter:Double, ; the typical/nominal size of the bga balls
+              pad-type:BGAPadType):
+  bga-pad(typical-ball-diameter, 
+          pad-type,
+          clearance(current-rules(), SolderMaskRegistration),
+          DENSITY-LEVEL)
+
+; BGAPadNames is a helper type to create a lookup table of precomputed
+; pad names given BGA parameters
+public deftype BGAPadNames
+
+; Lookup the pad name given row and column indices
+public defmulti get (b:BGAPadNames, r:Int, c:Int) -> Ref
+
+; Compute the pad names
+public defn bga-pad-names (num-rows:Int, num-cols:Int, row-chars-omitted?:False|String):
+  ; pick the omitted characters if not supplied fallback to a default
+  val row-chars-omitted = match(row-chars-omitted?):
+    (s:String): s
+    (_:False): "IOQSXZ"
+    
+  ; A small table of characters, each corresponds to the base character for a row
+  ; at a given index
+  val row-chars = to-tuple $
+    for ch in "ABCDEFGHIJKLMNOPQRSTUVWXYZ" filter:
+      not contains?(upper-case(row-chars-omitted), ch)
+  
+  ; helper function to create pad names given indices, follows the convention
+  ; "A" "B" "C" ... "AA" "AB" "AC" ... "BA" "BB" "BC" ... etc
+  val buf = StringBuffer()
+  defn row-name (idx:Int):
+    clear(buf)
+    var count = to-seq(0 to false)
+    let loop (row: Int = idx):
+      fatal("exceeded recursion depth when precomputing bga pad names") when next(count) >= 100
```

Comment:
No need for this check. If the algorithm is correct, this will never happen for 32-bit integers.